File: 347-top-k-frequent-elements/top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        vector<vector<int>> bucket(nums.size() + 1);
        unordered_map<int, int> freqMap;

        for (int num : nums) {
            freqMap[num]++;
        }

        for (auto i : freqMap) {
            int freq = i.second;
            bucket[i.second].push_back(i.first);
        }
        vector<int> res;

        for (int i = nums.size(); i>= 0 && res.size() < k; i--) {
            int currIndex = 0;
            while (currIndex < bucket[i].size() && res.size() < k) {
                res.push_back(bucket[i][currIndex]);
                currIndex++;
            }
        }
        return res;
    }
};

File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> mapFreq = new HashMap<>();
        for (int i : nums) {
            int currFreq = mapFreq.getOrDefault(i, 0);
            mapFreq.put(i, currFreq + 1);
        }
        PriorityQueue<int[]> pq  = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        for (int num : mapFreq.keySet()) {
            pq.add(new int[]{mapFreq.get(num), num});
            if (pq.size() > k) pq.poll();
        }
        int[] res = new int[k];
        int i = 0;
        while (!pq.isEmpty()) {
            res[i++] = pq.poll()[1]; 
        }
        return res;
    }
}